File: Javascript/Algorithms/QuickSort.js
var Qs = function(arr, start, end) {
    if (start < end) {
        var index = partition(arr, start, end);
        Qs(arr, start, index - 1);
        Qs(arr, index + 1, end);
    }
}

function partition(arr, s, e) {
    var pivot = arr[e];
    var index = s;
    for (var i = s; i <= e; i++) {
        if (arr[i] <= pivot) {
            swap(arr, i, index);
            index++;
        }
    }
    return index;
}

function swap(arr, a, b) {
    var tmp = arr[b];
    arr[b] = arr[a];
    arr[a] = tmp;
}
var arr = [5,4,3,1,0,5,7,8,9,7,9];
Qs(arr, 0, arr.length - 1);
console.log(arr);

File: Ruby/legacy/Algorithms/quicksort.rb
def quicksort(array)
  return sort(array,0,array.length-1)
end

def sort(array,left,right)
  return [] if left>=right
  p_index = partition(array,left,right)
  sort(array,left,p_index-1)
  sort(array,p_index+1,right)
  return array
end
def partition(array,left,right)
  p_index = left
  pivot = array[right]
  (left...right).each do |index|
    if array[index] <=pivot
      array[index],array[p_index] = array[p_index],array[index]
      p_index = p_index+1
    end
  end
  array[right],array[p_index] = array[p_index],array[right]
  p_index
end

puts quicksort([5,4,7,1,2,5])
